azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -2144,6 +2144,64 @@ def test_sends_down(self, stage, mocker, op):
         assert stage.send_op_down.call_args == mocker.call(op)
 
 
+@pytest.mark.describe(
+    "CoordinateRequestAndResponseStage - OCCURRENCE: pending_response is deleted, but another response is received"
+)
+class TestCoordinateRequestAndResponseStageRequestDeletedPendingOp(
+    CoordinateRequestAndResponseStageTestConfig
+):
+    @pytest.fixture
+    def event(self, fake_uuid):
+        return pipeline_events_base.ResponseEvent(
+            request_id=fake_uuid, status_code=200, response_body="response body"
+        )
+
+    @pytest.fixture
+    def request_response_op(self, mocker):
+        return pipeline_ops_base.RequestAndResponseOperation(
+            request_type="some_request_type",
+            method="SOME_METHOD",
+            resource_location="some/resource/location",
+            request_body="some_request_body",
+            callback=mocker.MagicMock(),
+        )
+
+    @pytest.fixture
+    def stage(self, mocker, cls_type, init_kwargs, fake_uuid, request_response_op):
+        stage = cls_type(**init_kwargs)
+        stage.pipeline_root = pipeline_stages_base.PipelineRootStage(
+            pipeline_configuration=mocker.MagicMock()
+        )
+        stage.next = mocker.MagicMock(spec=pipeline_stages_base.PipelineStage)
+        stage.send_event_up = mocker.MagicMock()
+        mocker.spy(stage, "report_background_exception")
+        return stage
+
+    @pytest.mark.it(
+        "Does not raise an exception when a RequestOperation callback with an error after being called with success"
+    )
+    def test_does_not_raise_on_second_callback(
+        self, stage, request_response_op, arbitrary_exception
+    ):
+
+        assert len(stage.pending_responses) == 0
+        assert stage.next.run_op.call_count == 0
+        assert request_response_op.callback_stack[0].call_count == 0
+
+        stage.run_op(request_response_op)
+
+        assert len(stage.pending_responses) == 1
+        assert stage.next.run_op.call_count == 1
+        request_op = stage.next.run_op.call_args[0][0]
+        callback_count = request_response_op.callback_stack[0].call_count
+        assert callback_count == 0
+
+        request_op.complete(error=arbitrary_exception)
```

Comment:
>request_op.complete(error=arbitrary_exception) [](http://example.com/codeflow?start=8&length=46)

You're only calling `request_op.complete()` once. I don't think the code you added would even be called in this case. (you can verify this by putting a breakpoint on your new code in the debugger and seeing if it gets hit when you run this test.

Reply:
I agree, I don't believe this test creates the conditions by which the bug would previously have surfaced. I would like to discuss the actual cause of this bug before commenting much further.

Reply:
Unit test removed as a new class object and added to an existing class. Ensured it properly tests the conditions under which the bug occurs. Marking as resolved.